aikido_perception/src/AprilTagsModule.cpp

```diff
@@ -0,0 +1,170 @@
+/** 
+ * @file AprilTagsModule.cpp
+ * @author Shushman Choudhury
+ * @date Apr 14, 2016
+ * @brief The implementation of the April Tags detector. The function detectObjects
+ * waits for the next ROS message on the specified topic.
+ */
+
+#include <aikido/perception/AprilTagsModule.hpp>
+#include <aikido/perception/shape_conversions.hpp>
+#include "dart/common/Console.h"
+#include <visualization_msgs/Marker.h>
+#include <visualization_msgs/MarkerArray.h>
+#include <ros/topic.h>
+#include <ros/ros.h>
+#include <ros/console.h>
+#include <boost/make_shared.hpp>
+#include <Eigen/Geometry>
+#include <tf/LinearMath/Vector3.h>
+#include <tf/LinearMath/Quaternion.h>
+#include <stdexcept>
+
+namespace aikido{
+namespace perception{
+
+//================================================================================================================================
+
+AprilTagsModule::AprilTagsModule( ros::NodeHandle node, std::string markerTopic, std::shared_ptr<ConfigDataLoader> configData,
+								  const dart::common::ResourceRetrieverPtr& resourceRetriever,	
+								  std::string referenceFrameId, dart::dynamics::Frame* referenceLink):
+		mNode(node),
+		mMarkerTopic(markerTopic),
+		mConfigData(configData),
+		mResourceRetriever(resourceRetriever),
+		mReferenceFrameId(referenceFrameId),
+		mReferenceLink(referenceLink)
+{
+	mListener = new tf::TransformListener();
+}
+
+//================================================================================================================================
+void AprilTagsModule::detectObjects(std::vector<dart::dynamics::SkeletonPtr>& skeleton_list,double _timeout, ros::Time timestamp)
+{
+	bool transform_lookedup = false;
+	//Looks at all detected tags, looks up config file 
+	//Appends new skeletons to skeleton list
+	do{
+		visualization_msgs::MarkerArrayConstPtr marker_message
+	 			= ros::topic::waitForMessage<visualization_msgs::MarkerArray>(mMarkerTopic,mNode,ros::Duration(_timeout));
```

Comment:
This is the one change I have doubts about. The `waitForMessage` is used in the current `prpy` implementation of the `AprilTagsModule`. I think we also had a discussion a while ago that we did not want this to continuously subscribe to the topic but rather get the next message when called upon to do so. I assumed that this formulation achieved that. I could be wrong though? We should discuss in person.